src/main/scala/ignition/core/jobs/utils/SparkContextUtils.scala

```diff
@@ -194,5 +199,102 @@ object SparkContextUtils {
       else
         objectHadoopFile(paths, minimumPaths)
     }
+
+    def parallelTextFiles(paths: Seq[String], maxBytesPerPartition: Long): RDD[String] = {
+      val s3Paths = parallelListFiles(paths)
+      val buckets = buildBuckets(s3Paths, maxBytesPerPartition)
+      val files = buckets.flatMap(_.paths)
+
+      val s3Key = sc.hadoopConfiguration.get("fs.s3n.awsAccessKeyId")
+      val s3Secret = sc.hadoopConfiguration.get("fs.s3n.awsSecretAccessKey")
+
+      val partitionedFiles = sc.parallelize(files).map(file => file -> ()).partitionBy(createPartitioner(buckets))
+
+      partitionedFiles.mapPartitions { files =>
+        val conf = new Configuration()
+        conf.set("fs.s3n.awsAccessKeyId", s3Key)
+        conf.set("fs.s3n.awsSecretAccessKey", s3Secret)
+        val codecFactory = new CompressionCodecFactory(conf)
+        files.map { case (path, _) => path } flatMap { s3Path =>
+          val fileSystem = FileSystem.get(new java.net.URI(s3Path), conf)
+          val path = new Path(s3Path)
+          val inputStream = Option(codecFactory.getCodec(path)) match {
+            case Some(compression) => compression.createInputStream(fileSystem.open(path))
+            case None => fileSystem.open(path)
+          }
+          Source.fromInputStream(inputStream).getLines()
+        }
+      }
+    }
+
+    private def createPartitioner(buckets: Seq[Bucket]): Partitioner = {
+      val size = buckets.size
+      val partitions: Map[Any, Int] = buckets.zipWithIndex.flatMap { case (bucket, index) => bucket.paths.map(path => path -> index) }.toMap
+      new Partitioner {
+        override def numPartitions: Int = size
+        override def getPartition(key: Any): Int = partitions(key)
+      }
+    }
+
+    private def buildBuckets(files: Seq[S3File], maxBytesPerPartition: Long): Seq[Bucket] = {
+      val buckets = ArrayBuffer.empty[Bucket]
```

Comment:
Shouldn't buckets be a var since you're changing it with ``+=``

```scala
files.distinctBy(_.path).foldLeft(ArrayBuffer.empty[Bucket]) {
  case (buckets, file) if buckets.isEmpty() || buckets.tail.size + file.size > maxBytesPerPartition => 
    empty + Bucket(1, new ArrayBuffer(file))
  case (buckets, file) => 
    val last = buckets.last
    buckets.init + last.copy(last.size+file.size, last.paths + file)
}  
```